### PR TITLE
Process identities in order

### PR DESCRIPTION
**Pull request by georgwiese**

### CODE REVIEW

**Comment by georgwiese** [comment]
There is one failing test. I just checked that it can be fixed by executing the outer query last (as was previously the case) instead of first.

This is probably some edge case we don't need to worry about in the RISC-V context. In the pre-compile case, we don't want to run the outer query after the memory accesses, because processing it is needed to receive inputs.

File: executor/src/witgen/sequence_iterator.rs [hunk 1]
```diff
@@ -26,16 +26,11 @@ pub struct DefaultSequenceIterator {
...
-    /// The number of rounds for the current row delta.
-    /// If this number gets too large, we will assume that we're in an infinite loop and exit.
-    current_round_count: usize,
 
     /// The steps on which we made progress.
     progress_steps: Vec<SequenceStep>,
 }
 
-const MAX_ROUNDS_PER_ROW_DELTA: usize = 100;
-
```
Comment (qwang98) [inline executor/src/witgen/sequence_iterator.rs:38]: Why we removed this (together with `current_round_count`)? Not that I'm complaining as this caused Keccak witgen to not run before (more than 100 rounds needed). How to prevent a row delta from running forever now?

**Comment by qwang98** [comment]
How is this achieved? `"What this gets us is that an identity is only executed if no previous identity identity has made progress."`

The old algorithm appears to also achieve this? They are just different by the order prover query, outter query, and identities are processed?

File: executor/src/witgen/sequence_iterator.rs (hunk 1 above)
Comment (georgwiese) [reply executor/src/witgen/sequence_iterator.rs:38]: Just because it was annoying to keep in this prototype ;)

This thing is only there because it would be annoying to be stuck in an infinite loop. As long as that doesn't happen, we don't need a limit. And it would only happen if there was a bug.

**Comment by georgwiese** [comment]
This is more of a prototype rn, (hence still a draft PR 😉), but I added some more description to the issue! We'll also talk about the pre-compile experiments in today's meeting, that should also give more context why this was needed.

File: executor/src/witgen/sequence_iterator.rs (hunk 1 above)
Comment (georgwiese) [reply executor/src/witgen/sequence_iterator.rs:38]: But actually, I think we should just remove it!

File: executor/src/witgen/sequence_iterator.rs [hunk 2]
```diff
@@ -59,7 +53,7 @@ impl DefaultSequenceIterator {
     /// If we're not at the last identity in the current row, just moves to the next.
     /// Otherwise, starts with identity 0 and moves to the next row if no progress was made.
     fn update_state(&mut self) {
-        while !self.is_done() && !self.has_more_actions() {
```
Comment (georgwiese) [inline executor/src/witgen/sequence_iterator.rs:62]: This `while` was equivalent to an if: `has_more_actions()` computes `self.cur_action_index < last_action_index` where `last_action_index >= 0`, and `start_next_round()` sets `self.cur_action_index = -1;`

**Comment by georgwiese** [comment]
Actually, I'd propose to merge this as is. It unblocks the auto-precompiles and the performance downsides are only relevant in the very first block, and only in runtime witgen (JITed witgen already works like this PR), which will be replaced hopefully soon.

Also, I just ran the Keccak benchmark and saw no difference to main:
```
executor-benchmark/keccak
                        time:   [12.189 s 12.402 s 12.611 s]
                        change: [-2.2709% +1.1330% +4.6290%] (p = 0.56 > 0.05)
                        No change in performance detected.
```

**Review by chriseth — APPROVED** [review]
